DM-21009: Refactor Posix and S3 datastores to reduce code duplication

timj opened the pull request:
Also uses dataclasses rather than named tuple for database dict.

timj committed:
Standardize the prologue between S3 and Posix datastores

timj committed:
Refactor S3 and POSIX datastores to share code

Shared code moved to FileLikeDatastore.

timj committed:
Replace DatabaseDict named tuple with a data class

TallJimbo commented on python/lsst/daf/butler/datastores/fileLikeDatastore.py:
```diff
@@ -0,0 +1,565 @@
+    templates : `FileTemplates`
+        File templates that can be used by this `Datastore`.
+    name : `str`
+        Label associated with this Datastore.
```
I didn't think Sphinx actually knew how to handle an `Attributes` section like this.  I've (recently) been adding class attributes (with type annotations with no default values) that I expect to shadow with instance attributes, just for the purposes of hanging docs off them (e.g. https://github.com/lsst/daf_butler/blob/2dcc2143d8cb69677398048342e1a046ab0bd77d/python/lsst/daf/butler/core/dimensions/elements.py#L212).  I think that's what @jonathansick recommended to me at one point, and now that I've gotten in the habit of doing it with type annotations (so that the attributes of regular classes are documented the same way as the attributes of dataclasses), it's really growing on me.

timj replied:
This is not new code as such because it all came from PosixDatastore. I'm happy to tweak it though.

TallJimbo commented on python/lsst/daf/butler/datastores/fileLikeDatastore.py:
```diff
@@ -0,0 +1,565 @@
+    Parameters
+    ----------
+    config : `DatastoreConfig` or `str`
+        Configuration.
```
Where a `str` means the path to some configuration?

TallJimbo replied:
You're also missing some parameters that are present in the actual `__init__` signature.

TallJimbo commented on python/lsst/daf/butler/datastores/fileLikeDatastore.py:
```diff
@@ -0,0 +1,565 @@
+        If a keyword is explicitly defined in the supplied ``config`` it
+        will not be overridden by this method if ``overwrite`` is `False`.
+        This allows explicit values set in external configs to be retained.
+        """
```
Should/could this docstring be inherited from the base `Datastore` instead?

timj replied:
Yes. Probably. I'm not sure sphinx is picking those up yet. Again this code is a move and not new.

TallJimbo commented on python/lsst/daf/butler/datastores/fileLikeDatastore.py:
```diff
@@ -0,0 +1,565 @@
+        component : `str`
+            The component to be retrieved (can be `None`).
+        readStorageClass : `StorageClass`
+            The `StorageClass` of the dataset being read.
```
This is a lot of items to return in a tuple. A `dict`, a `SimpleNamespace`, or a new `dataclass` instance might be better, at least if this is going to be called in more than a couple of places.

timj replied:
I was torn. They grew organically during the refactor and of course it's a bit of a red flag when you have to split the line that is calling this method before the equals sign because there are so many variables coming back. I was shying away from a dataclass or some struct thing because it seemed like a lot of overhead for what is an internal method, but I'll take another look at it.

TallJimbo reviewed (approved):
Looks great; only minor comments.

I wish I could knock out refactors this fast  :slightly_smiling_face: .

timj committed:
Switch undo event from named tuple to dataclass

timj committed:
Use relative path to import from location within core

timj committed:
Reorganize class documentation of datastores and use type annotations

timj committed:
Add some type annotations to Config, DatabaseDict and Formatter

timj committed:
Ignore the mypy cache

timj committed:
Move some duplicated code out of in-memory datastore

This led to some rewriting of the "StoredItemInfo" classes
to allow file datastores and in-memory datastores to share code.

timj committed:
Be more explicit about record to info conversions

timj committed:
Clarify config docstring

timj committed:
Return a class rather than many items in a tuple

timj commented:
@TallJimbo I've done your fixes which led to me also refactoring in memory datastore and adding a bunch of type annotations. I think these sufficiently to warrant a quick bonus review if you wouldn't mind.

timj committed:
Remove top level __init__ and stop importing lsstimport

TallJimbo commented on python/lsst/daf/butler/core/datastore.py:
```diff
@@ -47,6 +48,14 @@ class DatastoreValidationError(ValidationError):
+    name: str
+    undoFunc: typing.Callable
+    args: tuple
+    kwargs: dict
```
I'd recommend using `__slots__` here, because there may be a lot of these.  (It's a real shame there's no way for `dataclass` to do that for you, but I've looked into it - apparently the decorator fires too late for anything it does in that department to matter).

timj replied:
It had not even crossed my mind that dataclasses did not use slots. I would have put money on them being specifically designed to be efficient.

I see much discussion at ericvsmith/dataclasses#28

TallJimbo commented on python/lsst/daf/butler/core/datastore.py:
```diff
@@ -63,19 +65,17 @@ class DatastoreTransaction:
-    Event = Event
+    Event: ClassVar[Type] = Event
+
+    parent: Optional['DatastoreTransaction']
```
At the risk of sending you down another rabbit hole, if you do `from __future__ import annotations`, you don't need to use a string here.  As a heads-up, if you do that, you'll need to bump the Travis Python version to 3.7.

TallJimbo commented on python/lsst/daf/butler/core/datastore.py:
```diff
@@ -63,19 +65,17 @@ class DatastoreTransaction:
         self._log = []
 
-    def registerUndo(self, name, undoFunc, *args, **kwargs):
+    def registerUndo(self, name: str, undoFunc: Callable, *args: Any, **kwargs: Any) -> None:
```
I have been leaving off the return type annotation for functions that always return `None`, but I don't have a strong opinion on that.  Should we establish a convention one way or the other?

timj replied:
I was following the examples in the mypy documentation. I think I would tend towards an explicit `None` indicating that we don't return anything.

TallJimbo commented on python/lsst/daf/butler/core/datastore.py:
```diff
@@ -92,7 +92,7 @@ def registerUndo(self, name, undoFunc, *args, **kwargs):
 
     @contextlib.contextmanager
-    def undoWith(self, name, undoFunc, *args, **kwargs):
+    def undoWith(self, name: str, undoFunc: Callable, *args: Any, **kwargs: Any) -> Generator:
```
Hmm.  I think this is technically correct, but I've been using the return type of the decorated function (i.e. `typing.ContextManager`) in this case, because I find it more useful for human readers of the code.  But I suppose if we ever did get to the point of running mypy, it'd have to be like this, i.e. the return type of the pre-decorated function?

timj replied:
I used Generator here because I ran mypy and it told me I was doing it wrong.

TallJimbo commented on python/lsst/daf/butler/datastores/inMemoryDatastore.py:
```diff
@@ -91,6 +87,15 @@ class InMemoryDatastore(Datastore):
+    storageClassFactory: StorageClassFactory
+    """Factory for creating storage class instances from name."""
+
+    datasets: Dict[int, object]
```
You've used `Any` instead of `object` elsewhere (as have I), though I think they are indeed equivalent here.

timj replied:
I was hoping that `object` would imply "must be an instance of something" but I see that Python thinks that pretty much everything is an object so it doesn't really matter.

TallJimbo commented on python/lsst/daf/butler/core/databaseDict.py:
```diff
@@ -42,24 +46,24 @@ def __getitem__(self, i):
 
     @classmethod
-    def fields(cls):
+    def fields(cls) -> typing.Sequence[str]:
```
-> just `Sequence`, since it's imported directly?

timj replied:
`Sequence` is already `collections.abc.Sequence` so I was assuming that's not the same as `typing.Sequence`.

timj replied:
Do we want to always use the short imports for `typing` and be explicit with `collections.abc` when using inheritance?  I think that might be clearer since type annotations will be used a lot more than `collections.abc` in the code.

timj replied:
I've changed it to stop using `collections.abc`

TallJimbo commented on python/lsst/daf/butler/core/formatter.py:
```diff
@@ -289,7 +288,7 @@ def getLookupKeys(self):
         return self._mappingFactory.getLookupKeys()
 
-    def getFormatterClassWithMatch(self, entity):
+    def getFormatterClassWithMatch(self, entity) -> Tuple[LookupKey, Type]:
```
A `typing.Union`-based alias for the `entity` arguments would be pretty nice, I think.

timj replied:
Yes. I was definitely going to do that at some point but that seemed to be a distraction too far for the moment. 😄

TallJimbo commented on python/lsst/daf/butler/core/formatter.py:
```diff
@@ -397,7 +396,7 @@ def getFormatter(self, entity, *args, **kwargs):
         return formatter
 
-    def registerFormatter(self, type_, formatter):
+    def registerFormatter(self, type_, formatter) -> None:
```
Forget the annotations on the arguments?

timj committed:
Add 2019 to COPYRIGHT year

TallJimbo commented on python/lsst/daf/butler/core/storedFileInfo.py:
```diff
@@ -19,87 +19,51 @@
+    This is an empty base class.  Datastore implementations are expected to
+    write their own subclasses.
     """
+    pass
```
Might want to do `__slots__ = ()` here so derived class that have a reason to be compact can be.

TallJimbo commented on python/lsst/daf/butler/core/storedFileInfo.py:
```diff
@@ -19,87 +19,51 @@
+                (inspect.isclass(self.formatter) and issubclass(self.formatter, Formatter)):
+            self.formatter = self.formatter.name()
+        else:
+            raise ValueError(f"Supplied formatter '{self.formatter}' is not a Formatter")
```
`TypeError` instead?

TallJimbo commented on python/lsst/daf/butler/core/storedFileInfo.py:
```diff
@@ -19,87 +19,51 @@
+            # We trust that this string refers to a Formatter
+            pass
+        elif isinstance(self.formatter, Formatter) or \
+                (inspect.isclass(self.formatter) and issubclass(self.formatter, Formatter)):
```
Why do we need the `inspect` check?  Does `issubclass` raise instead of returning `False` when given non-types?

timj replied:
Yes. Amazingly enough `issubclass` raises if it's given anything other than a class, rather than returning `False` (or checking whether the instance is equivalent).

TallJimbo commented on python/lsst/daf/butler/core/storedFileInfo.py:
```diff
@@ -19,87 +19,51 @@
-        return f'{type(self).__qualname__}(path="{self.path}", formatter="{self.formatter}"' \
-            f' size={self.size}, checksum="{self.checksum}", storageClass="{self.storageClass.name}")'
+    def __post_init__(self):
+        # This modification prevents the data class from being frozen.
```
You may be interested in https://github.com/lsst/daf_butler/commit/19981afea055fc72ffefa1c985f6be0d0f913a6f (feel free to cherry-pick to this branch if you are).

timj replied:
I don't think that helps. It is no different in behavior to me doing `@dataclass(frozen=True)` and it still prevents me from updating the `formatter`.

TallJimbo commented on python/lsst/daf/butler/datastores/genericDatastore.py:
```diff
@@ -0,0 +1,169 @@
+    """
+
+    records: MutableMapping
+    """Place to store internal records about datasets."""
```
If this is only used in the default implementation of `removeStoredItemInfo`, I think it might be cleaner overall to leave that abstract (like `addStoredItemInfo`) and let subclasses define `records` or not; it might avoid some code duplication today, but I don't think `DatabaseDict` will survive adding vectorized inserts of datasets, and that I think that means we shouldn't expect the records for `InMemoryDatastore` to continue to look like the records for other `Datastores`.

timj replied:
Are you saying that at some point there is going to be a `Datastore.putMany()` that will process multiple in-memory datasets?

genericDatastore now has implementations for add/get/removeStoredItemInfo so it's not just `add` that is using records in this file. If dict-like record store disappears from FileLikeDatastore then we can move them around at that point I think.

TallJimbo replied:
I don't have a use case for `putMany`, but we definitely have a use case for a vectorized version of `Datastore.ingest`.

TallJimbo reviewed (approved):
A few more comments, once again, none of them major.

We should talk sometime about the future of `DatabaseDict` and the relationship between `Registry` and `Datastore`.  I don't really have a clear plan yet, but I think some fairly significant changes will be needed to add vectorized inserts.

timj committed:
Enable slots for StoredDatastoreItemInfo classes and Event

timj committed:
Change exception to TypeError when supplied formatter is wrong type

timj committed:
Add missing type annotations to Formatter

timj committed:
In type annotation prefer Any over object

timj committed:
Clean up slots and abc for DatabaseDict

timj committed:
Put more information in error message indicating that a dataset is missing

timj committed:
Use keyword constructor for all parameters

This avoids any possible confusion

timj committed:
Adjust order of attributes in DatastoreRecord

For datastores the posix_datastore_record database table
had columns whose order was determined by the order of the
entries in the dict defining the data types associated
with entries in the NamedTuple.  I had expected them
to be defined based on the NamedTuple definition so wrote
the Dataclass to match.  In the new implementation the
types definitions and the field orders match because
they are all backed by the dataclass, but in order for
the branch to be compatible with previous gen3 installations
we change the order here to match what was going on
previously.

timj committed:
Clarify why butlerRoot can be in datastore name
